File: f00ale-cpp_py/py/p01.py
#!/usr/bin/env python3

from aoclib import *

nums = {
    'one': 1,
    'two': 2,
    'three': 3,
    'four': 4,
    'five': 5,
    'six': 6,
    'seven': 7,
    'eight': 8,
    'nine': 9,
}

def main():
    data = readdata()
    ans1 = 0
    ans2 = 0
    for l in data:
        l = l.strip()
        first = last = None
        for c in l:
            if c >= '0' and c <= '9':
                if not first:
                    first = c
                last = c
        if first and last:
            tmp = 10*int(first)+int(last)
        ans1 += tmp
    print(ans1)
    for l in data:
        l = l.strip()
        for idx in range(len(l)):
            for s,n in nums.items():
                if s == l[idx:idx+len(s)]:
                    l = l[:idx] + str(n) + l[idx+1:]
        first = last = None
        for c in l:
            if c >= '0' and c <= '9':
                if not first:
                    first = c
                last = c
        if first and last:
            tmp = 10*int(first)+int(last)
        ans2 += tmp
    return ans1, ans2

if __name__ == '__main__':
    checkans(*main())

File: f00ale-cpp_py/src/p01.cpp
#include "aoc.h"
#include <vector>
#include <algorithm>
#include <iostream>
#include <tuple>
#include <set>

std::tuple<std::string, std::string> p01(const std::string &input) {
    int64_t ans1 = 0;
    int64_t ans2 = 0;

    {
        static const std::vector<std::string> nums = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        std::string line;
        for (const auto c: input) {
            if ((c >= '0' && c <= '9') || (c >= 'a' && c <= 'z')) {
                line += c;
            } else if (c == '\n') {
                auto calc = [](const auto &str) {
                    int first = 0, last = 0;
                    for (auto d: str) {
                        if (d >= '0' && d <= '9') {
                            last = d - '0';
                            if (!first) first = last;
                        }
                    }
                    return first * 10 + last;
                };
                ans1 += calc(line);

                std::string::size_type pos = 0;
                int first = 0, last = 0;
                while((pos = line.find_first_of("123456789otfsen", pos)) != std::string::npos)
                {
                    int tmp = 0;
                    if(line[pos] >= '0' && line[pos] <= '9') {
                        tmp = line[pos] - '0';
                    } else {
                        int idx = 0;
                        for(auto & s : nums) {
                            idx++;
                            if(line.find(s, pos) == pos) {
                                tmp = idx;
                                break;
                            }
                        }
                    }

                    pos++;
                    if(tmp) {
                        last = tmp;
                        if(!first) first = last;
                    }
                }
                ans2 += first * 10 + last;

                line.clear();
            }
        }

    }

    return {std::to_string(ans1), std::to_string(ans2)};
}
